components/debug/debuglog.hpp

```diff
@@ -18,32 +18,64 @@ namespace Debug
     };
 
     extern Level CurrentDebugLevel;
+}
+
+class Log
+{
+    static std::mutex sLock;
 
-    class LogWrapper
+    std::unique_lock<std::mutex> mLock;
+public:
+    // Locks a global lock while the object is alive
+    Log(Debug::Level level) :
+    mLock(sLock),
+    mLevel(level),
+    mWritten(false)
     {
-        static std::mutex sLock;
-
-        std::unique_lock<std::mutex> mLock;
-    public:
-        // Locks a global lock while the object is alive
-        LogWrapper(Level level) : mLock(sLock) { std::cout << static_cast<unsigned char>(level); }
-
-        // Perfect forwarding wrappers to give the chain of objects to cout
-        template<typename T>
-        const LogWrapper& operator<<(T&& rhs) const
-        { std::cout << std::forward<T>(rhs); return *this; }
-        template<typename T>
-        const LogWrapper& operator<<(const T& rhs) const
-        { std::cout << std::forward<const T&>(rhs); return *this; }
-
-        ~LogWrapper() { std::cout << std::endl; }
-    };
-}
+    }
+
+    // Perfect forwarding wrappers to give the chain of objects to cout
+    template<typename T>
+    Log& operator<<(T&& rhs)
+    {
+        if (mLevel <= Debug::CurrentDebugLevel)
+        {
+            WriteDebugLevel();
+            std::cout << std::forward<T>(rhs);
+        }
 
-// If the current log level is high enough, a LogWrapper object is constructed that
-// takes the global log lock, and is passed along the << chain until the ; at the end
-// of the statement where it goes out of scope, where it's destructed and releases
-// the lock, thus ensuring only a single thread is writing to the log at a time.
-#define LOG(L) if(L <= Debug::CurrentDebugLevel) Debug::LogWrapper(L)
+        return *this;
+    }
+    template<typename T>
+    Log& operator<<(const T& rhs)
+    {
+        if (mLevel <= Debug::CurrentDebugLevel)
+        {
+            WriteDebugLevel();
+            std::cout << std::forward<const T&>(rhs);
+        }
+
+        return *this;
+    }
+
+    ~Log()
+    {
+        if (mWritten)
+            std::cout << std::endl;
+    }
+
+private:
+    Debug::Level mLevel;
+    bool mWritten;
+
+    void WriteDebugLevel()
+    {
+        if (!mWritten)
+        {
+            std::cout << static_cast<unsigned char>(mLevel);
+            mWritten = true;
+        }
```

Comment:
I don't think this is necessary, (re)writing the log level for every `<<` call and needing to track if something was written for the destructor. Just checking if the log level checks out prior to writing anything should be good enough.

Reply:
I do not understand what "Just checking if the log level checks out" means.

Reply:
Doing in the constructor:
```
if(mLevel <= Debug::CurrentDebugLevel)
    std::cout << mLevel;
```
For the destructor:
```
if (mLevel <= Debug::CurrentDebugLevel)
    std::cout << std::endl;
```
And for writing the objects:
```
if (mLevel <= Debug::CurrentDebugLevel)
    std::cout << std::forward<T>(rhs);
```
There's no need to keep writing the debug level for every object, nor any need to keep a track of a `mWritten` variable.

Reply:
Done.